rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+# Implementation Plan: Filtering and designating results with sensitive textual content
+
+## Reviewers
+
+- [ ] TBD
+- [ ] TBD
+
+## Project links
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/377)
+- [Project Proposal](./20230309-project_proposal_detecting_sensitive_textual_content.md)
+- [Milestone (pending)](#)
+- [Implementation Plan: Managing A Sensitive Terms List](https://github.com/WordPress/openverse/pull/911)
+
+Please refer to additional prior art linked in the project thread. The most
+significant technical prior art is this exploratory PR for creating a filtered
+index: https://github.com/WordPress/openverse-api/pull/1108
+
+## Expected outcomes
+
+- Results are filtered efficiently based on a list of sensitive terms.
+- API results include a call-out on the result objects designating individual
+  results that have sensitive textual content.
+
+## Tools and dependencies
+
+We will not need any new tools on the API side for this work.
+
+## Feature flags
+
+We should introduce a single "feature flag" on the API as an undocumented and
+unstable query parameter `unstable__include_sensitive_results`.
+
+## API versioning
+
+There are no API versioning concerns. We already filter sensitive content based
+on provider data, we'll just be expanding upon that existing feature by
+incorporating a new signal into the filter. I do not believe this requires a new
+API version prefix or distinctly new parameter.
+
+## Areas touched
+
+Implementing this plan will require making changes in the following areas:
+
+- API
+- Elasticsearch
+- Ingestion server
+- Airflow data refresh DAG factory
+
+## Terms
+
+- "Origin index": This is the index we have now that includes all documents.
+- "Filtered index": This is the index we will have after this implementation
+  plan is implemented. Its documents are derived from the origin index but
+  exclude any that have sensitive terms in their textual content.
+
+## Overview of implementation process
+
+In depth descriptions of each of these steps are covered below. Each step will
+be implemented as a single pull request including appropriate tests.
+
+1. Update the ingestion server with a new action to use the reindex API to
+   generate a new index based on the origin index that only includes documents
+   that do not have sensitive terms in the three textual content fields
+   (`title`, `description`, `tags.name`). This is the filtered index.
+   - Requires the following:
+     - SENSITIVE_TERMS injected as new environment variable. See
```

Comment:
This is a very small detail but I'm wondering if a file stored in S3 and it's URL in the environment variable would be a better option that loading the entire list of sensitive terms from an environment variable. It's not an immediate issue but if we include sensitive terms from multiple languages, the list might be several megabytes which might exceed some character and size limits.

Reply:
I had thought about doing it that way as well, and I don't remember why I landed on using an environment variable. Alternatively, we could even request the list directly from GitHub, as it will be a public repository storing them.

The difficulty this presents is for testing data that doesn't need to be stored in a public repository. We could either have an `inane-terms.txt` in the repository for testing or use `SENSITIVE_TERMS_URL` as the primary and `SENSITIVE_TERMS` as a fallback that would only be used in non-production environments.